### DISCUSSION TITLE
openPMD viewer: visualize AMR simulation

### DISCUSSION THREAD

**Discussion by kli-jfp** [BLAST-WarpX/warpx#4725] [Data Analysis] [answered]
Hi! I recently got into running WarpX with mesh refinement with multiple levels. I can view the individual levels separately, but not together. Is there a current solution for viewing them together in the same plot? What is the recommended way to visualize this? 

Also, is there a way to create complex mesh refinement shapes or can you only make boxes as of right now? For example, I've tried creating a refined torus shape in 3D, but the result is a box enclosing the torus.

Thanks!

**Comment by RemiLehe** [comment, marked as answer]
Thanks for the question.

Regarding visualizing the fields with MR, there are two options:
- Having WarpX write the data in [plotfile](https://warpx.readthedocs.io/en/latest/dataanalysis/formats.html#output-formats) format, and then using `yt` (see [here](https://warpx.readthedocs.io/en/latest/dataanalysis/yt.html)), which does have extensive support for MR visualization.
- Having WarpX write the data in [openPMD](https://warpx.readthedocs.io/en/latest/dataanalysis/formats.html#output-formats) format, and then using `openPMD-viewer`. In that particular case, you can use the fact that, for refined levels, the fields outside the MR patch are seen as NaN (https://github.com/openPMD/openPMD-viewer/pull/334)
So, if you call `get_field` with `plot=True`, and call the fields in the order going from low-level to high level, you will automatically see the right thing. (since, in matplotlib's `imshow`, NaNs are just transparent):

```
ts.get_field('E', 'x', plot=True)  # This only plots level 0
ts.get_field('E_lvl1', 'x', plot=True)
ts.get_field('E_lvl2', 'x', plot=True)
```

**Reply by kli-jfp** [reply]
Thanks Remi, that works great!

**Comment by RemiLehe** [comment]
Regarding refined refinement: thanks for reporting this. Did you use this feature, in order to create the patch:
https://github.com/ECP-WarpX/WarpX/pull/4299

One thing to keep in mind is that, by construction in WarpX/AMReX, the MR refinement patch consists of a collection of patches of size [amr.blocking_factor](https://warpx.readthedocs.io/en/latest/usage/workflows/domain_decomposition.html#more-general-method)`. In other words, the granularity of the MR patch shape is `amr.blocking_factor`. In the above-mentioned torus case, maybe reducing `amr.blocking_factor` could help?

**Reply by kli-jfp** [reply]
Yes, I used ref_patch_function(x,y,z). Thanks for the tip, I'll try different settings of blocking_factor and see if it works.
